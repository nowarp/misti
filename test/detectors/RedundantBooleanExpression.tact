contract Test {
    a: Int = 0;
    fun test() {
        if ((self.a == 1) || (self.a == 2)) {} // OK
        if ((self.a == 1) || (1 == self.a)) {} // OK: No AST comparison API available
        if ((self.a == 1) || (self.a != 1)) {} // OK
        if (((self.a == 1) && (self.a != 1)) || (self.a == 1)) {} // OK

        if ((self.a > 0) || (self.a > 0)) {} // Bad
        if ((self.a == 1) || (self.a != 0) || (self.a == 1)) {} // Bad
        if ((self.a == 1) && (self.a != 0) && (self.a == 1)) {} // Bad
    }
}

