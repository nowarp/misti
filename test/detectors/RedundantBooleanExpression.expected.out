[MEDIUM] RedundantBooleanExpression: Redundant boolean expression: found duplicate conditions
test/detectors/RedundantBooleanExpression.tact:9:13:
   8 | 
>  9 |         if ((self.a > 0) || (self.a > 0)) {} // Bad
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  10 |         if ((self.a == 1) || (self.a != 0) || (self.a == 1)) {} // Bad
Help: Remove duplicate conditions to simplify the expression.
See: https://nowarp.io/tools/misti/docs/detectors/RedundantBooleanExpression

[MEDIUM] RedundantBooleanExpression: Redundant boolean expression: found duplicate conditions
test/detectors/RedundantBooleanExpression.tact:10:13:
   9 |         if ((self.a > 0) || (self.a > 0)) {} // Bad
> 10 |         if ((self.a == 1) || (self.a != 0) || (self.a == 1)) {} // Bad
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  11 |         if ((self.a == 1) && (self.a != 0) && (self.a == 1)) {} // Bad
Help: Remove duplicate conditions to simplify the expression.
See: https://nowarp.io/tools/misti/docs/detectors/RedundantBooleanExpression

[MEDIUM] RedundantBooleanExpression: Redundant boolean expression: found duplicate conditions
test/detectors/RedundantBooleanExpression.tact:11:13:
  10 |         if ((self.a == 1) || (self.a != 0) || (self.a == 1)) {} // Bad
> 11 |         if ((self.a == 1) && (self.a != 0) && (self.a == 1)) {} // Bad
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  12 |     }
Help: Remove duplicate conditions to simplify the expression.
See: https://nowarp.io/tools/misti/docs/detectors/RedundantBooleanExpression